File: NEMbox/hotKey.py
import objc
from AppKit import NSApplication
from AppKit import NSKeyUp
from AppKit import NSSystemDefined
from PyObjCTools import AppHelper


class HotKeyApp(NSApplication):
    repeated = False

    @objc.python_method
    def setMusicMenu(self,menu):
        self.menu = menu

    def sendEvent_(self, event):
        if event.type() is NSSystemDefined and event.subtype() == 8:
            data = event.data1()
            key_code = (data & 0xFFFF0000) >> 16
            key_flags = data & 0x0000FFFF
            key_state = (key_flags & 0xFF00) >> 8
            key_repeat = key_flags & 0x1
        
        if key_state is NSKeyUp:
            if self.repeated:
                self.repeated = False
        elif key_code in (20, 18):
            self.menu.previous_song()
        elif key_code == 16:
            self.menu.pause()
        elif key_code in (19, 17):
            self.menu.next_song()

File: NEMbox/osx/hotkey.c
#include<stdio.h>
#import <Cocoa/Cocoa.h>
#import <Foundation/Foundation.h>
#import <AppKit/AppKit.h>
#import <IOKit/hidsystem/ev_keymap.h>

//Compile gcc -fobjc-arc -framework Cocoa -x objective-c -shared -o HotKey.lib hotkey.c
void (*play)(void);
void (*nextSong)(void);
void (*lastSong)(void);

void setContorllFunc(void (*playFunc)(void),void (*nextFunc)(void),void (*lastFunc)(void)){
    play = playFunc;
    nextSong = nextFunc;
    lastSong = lastFunc;
}

@interface HotKeyApp : NSApplication

@end

@implementation HotKeyApp
    - (void)sendEvent:(NSEvent*)anEvent{
        if([anEvent type] == NSEventTypeSystemDefined && [anEvent subtype] == 8){
            long data = [anEvent data1];
            int keyCode = ((data & 0xffff0000) >> 16);
            int keyFlags = (data & 0x0000ffff);
            int keyState = (((keyFlags & 0xff00) >> 8)) == 0xa;
            
            if(keyState){
                switch(keyCode){
                    case NX_KEYTYPE_PLAY:
                        play();
                        break;
                    
                    case NX_KEYTYPE_FAST:
                    case NX_KEYTYPE_NEXT:
                        nextSong();
                        break;
                    
                    case NX_KEYTYPE_REWIND:
                    case NX_KEYTYPE_PREVIOUS:
                        lastSong();
                        break;
                }
            }
            // return;
        }
	    // [super sendEvent:anEvent];
    }
@end
NSApplication* app;
void setup(){
    @autoreleasepool{
         app = [HotKeyApp sharedApplication];
        [app setActivationPolicy:NSApplicationActivationPolicyProhibited];
        [app activateIgnoringOtherApps:YES];
        [NSApp run];
    }
}

void stop(){
    @autoreleasepool{
        [NSApp terminate:app];
    }
}